pitop/pma/ultrasonic_sensor.py

```diff
@@ -1,61 +1,47 @@
-from pitopcommon.logger import PTLogger
-
 from pitop.core.mixins import (
     Stateful,
     Recreatable,
 )
-from pitop.pma.common import get_pin_for_port
-
-
-# Modified version of gpiozero's DistanceSensor class that only uses 1 pin
-#
-# Note: all private member variables are semi-private to follow upstream gpiozero convention
-# and to override inherited functions
-class UltrasonicSensor(Stateful, Recreatable, SmoothedInputDevice):
-    ECHO_LOCK = Lock()
-
-    def __init__(
-        self,
-        port_name,
-        queue_len=9,
-        max_distance=3,
-        threshold_distance=0.3,
-        partial=False,
-        name="ultrasonic"
-    ):
+from pitop.pma.common.utils import Port
+from .ultrasonic_sensor_base import UltrasonicSensorMCU, UltrasonicSensorRPI
+
+
+valid_analog_ports = ["A1", "A3"]
+
+
+class UltrasonicSensor(Stateful, Recreatable):
+    def __init__(self,
+                 port_name,
+                 queue_len=5,
+                 max_distance=3,
+                 threshold_distance=0.3,
+                 partial=False,
+                 name="ultrasonic"
+                 ):
 
+        assert (port_name in Port)
         self._pma_port = port_name
         self.name = name
 
-        SmoothedInputDevice.__init__(self,
-                                     get_pin_for_port(self._pma_port),
-                                     pull_up=False,
-                                     queue_len=queue_len,
-                                     sample_wait=0.06,
-                                     partial=partial,
-                                     ignore=frozenset({None}),
-                                     )
-
-        try:
-            if max_distance <= 0:
-                raise ValueError('invalid maximum distance (must be positive)')
-            self._max_distance = max_distance
-            self.threshold = threshold_distance / max_distance
-            self.speed_of_sound = 343.26  # m/s
-
-            self._echo = Event()
-            self._echo_rise = None
-            self._echo_fall = None
-            self.pin.edges = 'both'
-            self.pin.bounce = None
-            self.pin.when_changed = self._echo_changed
-            self._queue.start()
-        except Exception:
-            self.close()
-            raise
+        if port_name in valid_analog_ports:
+            self.__ultrasonic_device = UltrasonicSensorMCU(port_name=port_name,
+                                                           queue_len=queue_len,
+                                                           max_distance=max_distance,
+                                                           threshold_distance=threshold_distance,
+                                                           partial=partial,
+                                                           name=name
+                                                           )
+        else:
+            self.__ultrasonic_device = UltrasonicSensorRPI(port_name=port_name,
```

Comment:
As an alternative to this or a factory method, I think it would be possible to dynamically inherit from these classes with the use of a [metaclass ](https://stackoverflow.com/a/100146). On the other hand this is the pattern used to manage the multiple implementations of the Camera component also and is less confusing.

Reply:
Possibly yeah, I read [this](https://realpython.com/factory-method-python/) when researching how to do this - weird they don't mention using a metaclass as it does seem ideal. I also can't find many examples of using a metaclass for this purpose but I'll definitely try to learn more about it. The dynamic inheritance does bother me in terms of documentation and general readability for users though, but if this was being done from scratch (as opposed to a bolt-on fix) I'd probably explore these more exotic options

Reply:
I agree a metaclass would be a good way to do this, it would also remove a lot of the code related to accessing properties on the underlying Ultrasonic device and enforce that the base classes need the same API. @PiTopRyan there is a really good overview on stackoverflow [here](https://stackoverflow.com/a/6581949)... someone reallllllllly wanted those upvotes 😆.

Reply:
> Metaclasses are deeper magic that 99% of users should never worry about it. If you wonder whether you need them, you don't (the people who actually need them to know with certainty that they need them and don't need an explanation about why).

Classic 😄 Great explanation on that post though!

Reply:
Just to clarify, I think a metaclass would be appropriate here but it wouldn't make a big enough difference to prevent merging 😄 Maybe a comment in the ultrasonic_sensor_base file saying the two underlying classes should be kept as similar as possible to this class would be good, but even then only if you can be bothered.